src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.cs

```diff
@@ -0,0 +1,669 @@
+            }
+
+            var matchingItems = itemsInList.Where(r => r.MatchedFilterText)
+                                           .Select(t => GetOrCreateRoslynItem(t.CompletionItem))
+                                           .AsImmutable();
+
+            var chosenItems = completionService.FilterItems(document, matchingItems, filterText);
+
+            var recentItems = _recentItems;
+
+            // Of the items the service returned, pick the one most recently committed
+            var bestItem = GetBestItemBasedOnMRU(chosenItems, recentItems);
+            var updatedFilters = GetUpdatedFilters(sortedList, itemsInList, filters, filterText);
+            VSCompletionItem uniqueItem = null;
+            int selectedItemIndex = 0;
+
+            // TODO: Can we get away with less complexity here by only doing hard select on preselection and not on regular filter text matching / etc...
+            // https://github.com/dotnet/roslyn/issues/29108
+
+            // Determine if we should consider this item 'unique' or not.  A unique item
+            // will be automatically committed if the user hits the 'invoke completion' 
+            // without bringing up the completion list.  An item is unique if it was the
+            // only item to match the text typed so far, and there was at least some text
+            // typed.  i.e.  if we have "Console.$$" we don't want to commit something
+            // like "WriteLine" since no filter text has actually been provided.  HOwever,
+            // if "Console.WriteL$$" is typed, then we do want "WriteLine" to be committed.
+            if (bestItem != null)
+            {
+                selectedItemIndex = itemsInList.IndexOf(i => Equals(GetOrCreateRoslynItem(i.CompletionItem), bestItem));
+                if (selectedItemIndex > -1 && bestItem != null && matchingItems.Length == 1 && filterText.Length > 0)
+                {
+                    uniqueItem = highlightedList[selectedItemIndex].CompletionItem;
+                }
+            }
+
+            // If we don't have a best completion item yet, then pick the first item from the list.
+            var bestOrFirstCompletionItem = bestItem ?? GetOrCreateRoslynItem(itemsInList.First().CompletionItem);
+
+            var updateSelectionHint = GetUpdateSelectionHint(bestOrFirstCompletionItem, snapshot, caretPosition, triggerReason, filterText, selectedItemIndex, hasSuggestedItemOptions);
+
+            if (selectedItemIndex == -1)
+            {
+                selectedItemIndex = 0;
+            }
+
+            return new AsyncCompletionData.FilteredCompletionModel(
+                highlightedList, selectedItemIndex, updatedFilters,
+                updateSelectionHint, centerSelection: true, uniqueItem);
+        }
+
+        private RoslynCompletionItem GetBestItemBasedOnMRU(ImmutableArray<RoslynCompletionItem> chosenItems, ImmutableArray<string> recentItems)
```

Comment:
all this code (including code above and below in this file) appears to have been copied from our existing feature.  I cannot figure out why we would copy that code instead of just having all that logic in a *single* location that both the old and new completoin systems can call into.

I basically have nearly no way to validate this is correct.  And it's enormously painful to review.  Ineed to bring up both files and go over them line by line.  I fundamentally do not understand why this has been done this way.  It just opens up the opportunity to have bugs and to have different behaviors between the two.

I'm skipping reviewing the rest of this file until we can come up witha  good way to review all this duplication.